opm/simulators/flow/ReservoirCouplingMaster.cpp

```diff
@@ -44,41 +44,133 @@ ReservoirCouplingMaster::ReservoirCouplingMaster(
 ) :
     comm_{comm},
     schedule_{schedule}
-{ }
+{
+    this->start_date_ = this->schedule_.getStartTime();
+}
 
 // ------------------
 // Public methods
 // ------------------
 
+double ReservoirCouplingMaster::maybeChopSubStep(
+        double suggested_timestep_original, double elapsed_time
+) const
+{
+    // Check if the suggested timestep needs to be adjusted based on the slave processes'
+    // next report step, or if the slave process has not started yet: the start of a slave process.
+    double start_date = static_cast<double>(this->start_date_);
+    TimePoint step_start_date{start_date + elapsed_time};
+    TimePoint step_end_date{step_start_date + suggested_timestep_original};
+    TimePoint suggested_timestep{suggested_timestep_original};
+    for (unsigned int i = 0; i < this->num_slaves_; i++) {
+        double slave_start_date = static_cast<double>(this->slave_start_dates_[i]);
+        TimePoint slave_next_report_date{this->slave_next_report_time_offsets_[i] + slave_start_date};
+        if (slave_start_date > step_end_date) {
+            // The slave process has not started yet, and will not start during this timestep
+            continue;
+        }
+        TimePoint slave_elapsed_time;
+        if (slave_start_date <= step_start_date) {
+            // The slave process has already started, and will continue during this timestep
+            if (slave_next_report_date > step_end_date) {
+                // The slave process will not report during this timestep
+                continue;
+            }
+            // The slave process will report during this timestep
+            slave_elapsed_time = slave_next_report_date - step_start_date;
+        }
+        else {
+            // The slave process will start during the timestep, but not at the beginning
+            slave_elapsed_time = slave_start_date - step_start_date;
+        }
+        suggested_timestep = slave_elapsed_time;
+        step_end_date = step_start_date + suggested_timestep;
```

Comment:
We update `step_end_date` for each iteration of the loop if the current value is less than the previous value (since if it is not less than that value, the execution will not reach this point). This is to compute the minimum `step_end_date` and the corresponding `suggested_time_step` over all the slaves. So it is not necessarily "the last one that wins", but rather the one that implies the smallest `step_end_date`. I have added a comment in the code to explain this. See the last commit.